File: desktop/management_credentials_test.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"path/filepath"
	"strings"
	"testing"

	"fairy/app/core"
	"fairy/app/edge"
	"fairy/runtime/config"
	"fairy/runtime/observability"
	api "fairy/transport/web"
)

func TestManagementFrontendBoundPayloadsNeverEchoCredentials(t *testing.T) {
	secrets := []string{
		"sk-live-fixture-secret-value",
		"sf-semantic-fixture-secret",
		"search-provider-fixture-secret",
		"plugin-credential-fixture-secret",
		"seekdb-private-password",
		"qq-access-token-fixture",
	}
	modelSecret := secrets[0]
	semanticSecret := secrets[1]
	store := observability.NewLogStore(8)
	store.Append(observability.EntryInput{
		Level:   "error",
		Logger:  "companion",
		Message: "request Authorization: Bearer " + modelSecret,
		Fields: []observability.FieldInput{
			{Key: "apiKey", Value: modelSecret},
			{Key: "password", Value: secrets[4]},
			{Key: "access_token", Value: secrets[5]},
		},
	})
	host := scriptedManagement{
		overview: func(context.Context) (edge.Overview, error) {
			return edge.Overview{
				Bootstrap: core.BootstrapStatus{AppName: "FAIRY", CoreVersion: "dev"},
				Storage:   api.StorageStatus{Ready: false, Mode: "production", Storage: "seekdb", Error: "dial failed with [seekdb-credential]"},
				SecretKey: edge.SecretKeyStatus{Ready: true, Mode: "production"},
				Model:     edge.ModelStatus{Configured: true, Protocol: "openai_compatible_api", Model: "test", AuthMode: "bearer_key"},
				Semantic:  edge.SemanticStatus{Provider: "siliconflow", Enabled: true, Configured: true, CredentialConfigured: true},
				WebSearch: config.WebSearchStatus{Enabled: true, Ready: true},
			}, nil
		},
		saveModel: func(write edge.ModelWrite) (edge.ModelStatus, error) {
			if write.APIKey != modelSecret {
				t.Fatalf("model host received %q", write.APIKey)
			}
			return edge.ModelStatus{Configured: true, Protocol: write.Protocol, Model: write.Model, AuthMode: write.AuthMode}, nil
		},
		model: func() (edge.ModelStatus, error) {
			return edge.ModelStatus{Configured: true, Protocol: "openai_compatible_api", Model: "test", AuthMode: "bearer_key"}, nil
		},
		saveSemantic: func(write edge.SemanticWrite) (edge.SemanticStatus, error) {
			if write.APIKey != semanticSecret {
				t.Fatalf("semantic host received %q", write.APIKey)
			}
			return edge.SemanticStatus{Provider: write.Provider, Enabled: true, Configured: true, CredentialConfigured: true}, nil
		},
		semantic: func() (edge.SemanticStatus, error) {
			return edge.SemanticStatus{Provider: "siliconflow", Enabled: true, Configured: true, CredentialConfigured: true}, nil
		},
		qq: func() (edge.QQSettings, error) {
			return edge.QQSettings{SchemaVersion: 1, GroupAllowlist: []string{"123"}}, nil
		},
		saveQQ: func(settings edge.QQSettings) (edge.QQSettings, error) {
			return settings, nil
		},
		plugins: func() (edge.PluginStatus, error) {
			return edge.PluginStatus{}, edge.ErrPluginHostUnavailable
		},
		logs: func(filter edge.LogFilter) (edge.LogSnapshot, error) {
			return store.Query(filter), nil
		},
		trace: func(_ context.Context, traceID string) (edge.TraceDetail, error) {
			return edge.TraceDetail{
				TraceID: traceID, Status: "ok", DurationMS: 12,
				Spans: []observability.TraceSpan{{SpanID: "span-1", Operation: "model", Status: "ok", DurationMS: 12, Attributes: map[string]string{"status": "ok"}}},
			}, nil
		},
	}
	service := NewCoreService()
	service.edge = &fakeOwnedRuntime{host: host}

	modelStatus, err := service.SaveManagementModel(edge.ModelWrite{
		ModelConnectionInput: config.ModelConnectionInput{Protocol: "openai_compatible_api", Model: "test", AuthMode: "bearer_key"},
		APIKey:               modelSecret,
	})
	if err != nil {
		t.Fatal(err)
	}
	semanticStatus, err := service.SaveManagementSemantic(edge.SemanticWrite{Provider: "siliconflow", Enabled: true, APIKey: semanticSecret})
	if err != nil {
		t.Fatal(err)
	}
	overview, err := service.ManagementOverview()
	if err != nil {
		t.Fatal(err)
	}
	qq, err := service.SaveManagementQQ(edge.QQSettings{SchemaVersion: 1, GroupAllowlist: []string{"123"}})
	if err != nil {
		t.Fatal(err)
	}
	logs, err := service.ManagementLogs()
	if err != nil {
		t.Fatal(err)
	}
	if len(logs.Entries) != 1 {
		t.Fatalf("logs = %#v", logs)
	}
	trace, err := service.ManagementTrace("trace-1")
	if err != nil {
		t.Fatal(err)
	}
	_, pluginErr := service.ManagementPlugins()
	if !errors.Is(pluginErr, edge.ErrPluginHostUnavailable) {
		t.Fatalf("ManagementPlugins() error = %v", pluginErr)
	}

	payloads := []any{modelStatus, semanticStatus, overview, qq, logs, trace, pluginErr.Error()}
	for _, payload := range payloads {
		assertNoSecretPlaintext(t, payload, secrets...)
	}
	if strings.Contains(logs.Entries[0].Message, modelSecret) {
		t.Fatalf("log message was not redacted: %#v", logs.Entries[0])
	}
	for _, field := range logs.Entries[0].Fields {
		if field.Value != observability.RedactedValue {
			t.Fatalf("log field %q = %q, want redacted", field.Key, field.Value)
		}
	}
}

func TestManagementSaveErrorsDoNotEchoSubmittedCredentials(t *testing.T) {
	secret := "sk-live-rejected-secret"
	host := scriptedManagement{
		saveModel: func(edge.ModelWrite) (edge.ModelStatus, error) {
			return edge.ModelStatus{}, errors.New("model credential is required")
		},
		saveSemantic: func(edge.SemanticWrite) (edge.SemanticStatus, error) {
			return edge.SemanticStatus{}, errors.New("semantic credential is required")
		},
	}
	service := NewCoreService()
	service.edge = &fakeOwnedRuntime{host: host}
	_, modelErr := service.SaveManagementModel(edge.ModelWrite{APIKey: secret})
	if modelErr == nil {
		t.Fatal("SaveManagementModel() error = nil")
	}
	_, semanticErr := service.SaveManagementSemantic(edge.SemanticWrite{APIKey: secret})
	if semanticErr == nil {
		t.Fatal("SaveManagementSemantic() error = nil")
	}
	assertNoSecretPlaintext(t, modelErr.Error(), secret)
	assertNoSecretPlaintext(t, semanticErr.Error(), secret)
}

func assertNoSecretPlaintext(t *testing.T, value any, secrets ...string) {
	t.Helper()
	encoded, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	text := string(encoded)
	for _, secret := range secrets {
		if strings.Contains(text, secret) {
			t.Fatalf("payload contained %q: %s", secret, text)
		}
	}
}

func TestDesktopFrontendProductionSourcesOmitCredentialPlaintext(t *testing.T) {
	tokens := []string{
		"FAIRY_API_TOKEN",
		"127.0.0.1:8787",
		"fairy.apiToken",
		"postgres://",
		"mysql://",
		"sk-live-",
		"qq-access-token",
		"seekdb-private-password",
	}
	err := walkDesktopProductionArtifacts(func(path string, content []byte) error {
		if !strings.Contains(path, string(filepath.Separator)+"frontend"+string(filepath.Separator)) {
			return nil
		}
		if strings.Contains(filepath.Base(path), ".test.") {
			return nil
		}
		text := string(content)
		lower := strings.ToLower(text)
		for _, token := range tokens {
			if strings.Contains(text, token) {
				t.Errorf("%s contains forbidden credential marker %q", path, token)
			}
		}
		if strings.Contains(lower, "pmhq") {
			t.Errorf("%s contains PMHQ", path)
		}
		if strings.Contains(text, "Authorization") {
			t.Errorf("%s contains Authorization", path)
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
}

File: desktop/frontend/src/managementCredentials.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { readFileSync, readdirSync, statSync } from "node:fs";
import { extname, join } from "node:path";
import { fileURLToPath } from "node:url";

const frontendRoot = fileURLToPath(new URL("./", import.meta.url));
const bindingsRoot = fileURLToPath(new URL("../bindings", import.meta.url));

const forbiddenSourceTokens = [
  "FAIRY_API_TOKEN",
  "127.0.0.1:8787",
  "fairy.apiToken",
  "Authorization",
  "postgres://",
  "mysql://",
  "sk-live-",
];

function walkFiles(root, files = []) {
  for (const name of readdirSync(root)) {
    const path = join(root, name);
    const info = statSync(path);
    if (info.isDirectory()) {
      if (name === "node_modules") continue;
      walkFiles(path, files);
      continue;
    }
    if (name.includes(".test.")) continue;
    if ([".js", ".jsx", ".mjs", ".css"].includes(extname(name))) files.push(path);
  }
  return files;
}

function renderManagementDOM(fixtures) {
  return [
    `应用 ${fixtures.overview.bootstrap.appName} ${fixtures.overview.bootstrap.coreVersion}`,
    `存储 ${fixtures.overview.storage.storage} ${fixtures.overview.storage.ready ? "就绪" : "未就绪"}`,
    `密钥 ${fixtures.overview.secretKey.ready ? "已配置" : "未就绪"}`,
    `模型 ${fixtures.overview.model.configured ? `${fixtures.overview.model.model}（已配置）` : "未配置"}`,
    `语义检索 ${fixtures.overview.semanticEmbedding.configured ? "已配置" : "未配置"}`,
    `诊断 ${fixtures.overview.storage.error}`,
    `模型凭据 ${fixtures.modelStatus.configured ? "已配置" : "未配置"}`,
    `语义凭据 ${fixtures.semanticStatus.credentialConfigured ? "已配置" : "未配置"}`,
    fixtures.saveError.message,
    fixtures.pluginError.message,
    fixtures.qqSettings.groupAllowlist.join("\n"),
    `搜索 ${fixtures.webSearch.ready ? "就绪" : "未就绪"}`,
    fixtures.logEntry.message,
    `${fixtures.trace.spans[0].operation} · ${fixtures.trace.spans[0].status} · ${fixtures.trace.spans[0].durationMs}ms`,
    `路径 ${fixtures.backup.path}`,
    `文件数 ${fixtures.backup.fileCount}`,
  ].join("\n");
}

test("management production sources do not embed Core bearer, remote endpoint, or secret material", () => {
  const files = walkFiles(frontendRoot).concat(walkFiles(bindingsRoot));
  assert.ok(files.length > 0);
  for (const file of files) {
    const text = readFileSync(file, "utf8");
    for (const token of forbiddenSourceTokens) {
      assert.doesNotMatch(text, new RegExp(token.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")), `${file} contains ${token}`);
    }
    assert.doesNotMatch(text, /pmhq/i, `${file} contains PMHQ`);
    assert.doesNotMatch(text, /Bearer\s+\S+/);
  }
});

test("management DOM, log, toast, and error fixtures never echo credentials", () => {
  const secrets = {
    model: "sk-live-fixture-secret-value",
    semantic: "sf-semantic-fixture-secret",
    search: "search-provider-fixture-secret",
    plugin: "plugin-credential-fixture-secret",
    database: "seekdb-private-password",
    qq: "qq-access-token-fixture",
  };
  const fixtures = {
    overview: {
      bootstrap: { appName: "FAIRY", coreVersion: "dev" },
      secretKey: { ready: true, mode: "production" },
      model: { configured: true, model: "test", authMode: "bearer_key" },
      semanticEmbedding: { configured: true, credentialConfigured: true, provider: "siliconflow" },
      storage: { ready: false, storage: "seekdb", mode: "production", error: "dial failed with [seekdb-credential]" },
    },
    modelStatus: { configured: true, protocol: "openai_compatible_api", model: "test", authMode: "bearer_key", secretStorageMigrated: true },
    semanticStatus: { provider: "siliconflow", enabled: true, configured: true, credentialConfigured: true },
    webSearch: { enabled: true, ready: true },
    qqSettings: { schemaVersion: 1, groupAllowlist: ["123"] },
    pluginError: { message: "plugin host is not configured" },
    logEntry: { sequence: 1, level: "info", message: "runtime ready", fields: [{ key: "authorization", value: "[REDACTED]" }] },
    saveError: { message: "model credential is required" },
    backup: { path: "/tmp/fairy-backup", fileCount: 1, createdAtUnixMs: 1 },
    workspace: { section: "tracing", traceId: "trace-1", messageId: "message-1", logLevel: "warn", pluginInstanceId: "echo-1" },
    trace: { traceId: "trace-1", spans: [{ spanId: "span-1", operation: "model", status: "ok", durationMs: 12 }] },
    clearedForm: { protocol: "openai_compatible_api", endpoint: "", model: "test", apiKey: "" },
  };
  const rendered = renderManagementDOM(fixtures);
  const payload = JSON.stringify(fixtures) + rendered;
  for (const secret of Object.values(secrets)) {
    assert.doesNotMatch(payload, new RegExp(secret.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")));
  }
  assert.doesNotMatch(payload, /Bearer\s+\S+/);
  assert.doesNotMatch(payload, /FAIRY_API_TOKEN|127\.0\.0\.1:8787|fairy\.apiToken|postgres:\/\/|mysql:\/\//);
  assert.doesNotMatch(payload, /"apiKey"\s*:\s*"[^"]+"/);
  assert.match(JSON.stringify(fixtures.clearedForm), /"apiKey":""/);
  assert.match(rendered, /模型凭据 已配置/);
  assert.match(rendered, /\[seekdb-credential\]/);
});

test("model save handler clears the password field and never logs or toasts the credential", () => {
  const source = readFileSync(new URL("./management.jsx", import.meta.url), "utf8");
  assert.match(source, /SaveManagementModel\(/);
  assert.match(source, /setForm\(\(current\) => \(\{ \.\.\.current, apiKey: "" \}\)\)/);
  assert.match(source, /type="password"/);
  assert.match(source, /data\?\.configured \? "已配置" : "未配置"/);
  assert.match(source, /semantic\.data\?\.credentialConfigured \? "已配置" : "未配置"/);
  assert.match(source, /function hostError\(err\) \{\n  if \(err && typeof err\.message === "string"/);
  assert.match(source, /<span>\{entry\.message\}<\/span>/);
  assert.doesNotMatch(source, /entry\.fields/);
  assert.doesNotMatch(source, /JSON\.stringify\((?:err|error|form|entry|data)\)/);
  assert.doesNotMatch(source, /console\.(?:log|debug|info|warn|error)\([^)]*apiKey/);
  assert.doesNotMatch(source, /setActionError\(form\.apiKey\)/);
  assert.doesNotMatch(source, /data\.storage\.(?:password|user|dsn)/);
  assert.doesNotMatch(source, /webSearch\.baseUrl/);
});
